AppCenterIdentity/AppCenterIdentity/Internals/Storage/MSKeychainAuthTokenStorage.m

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import "MSKeychainAuthTokenStorage.h"
+#import "MSAuthTokenInfo.h"
+#import "MSIdentityConstants.h"
+#import "MSIdentityPrivate.h"
+#import "MSKeychainUtil.h"
+#import "MSLogger.h"
+#import "MSUtility.h"
+
+@implementation MSKeychainAuthTokenStorage
+
+- (nullable NSString *)retrieveAuthToken {
+  NSMutableArray<MSAuthTokenInfo *> *authTokensHistory = [self authTokensHistoryState];
+  MSAuthTokenInfo *latestAuthTokenInfo = authTokensHistory.lastObject;
+  return latestAuthTokenInfo.authToken;
+}
+
+- (nullable NSString *)retrieveAccountId {
+  return [MS_USER_DEFAULTS objectForKey:kMSIdentityMSALAccountHomeAccountKey];
+}
+
+// TODO: Finish the implementation, this method will be used to get logs from DB for a period when token was active.
+- (MSAuthTokenInfo *)oldestAuthToken {
+  return [MSAuthTokenInfo new];
+
+  /*
+  // Read token array from storage.
+  NSMutableArray<MSAuthTokenInfo *> *tokenArray = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
+  if (!tokenArray || tokenArray.count == 0) {
+    return nil;
+  }
+
+  MSAuthTokenInfo *authTokenInfo = tokenArray.firstObject;
+  NSDate *nextChangeTime = tokenArray.count > 1 ? tokenArray[1].startTime : nil;
+  if ([authTokenInfo.endTime laterDate:nextChangeTime]) {
+    return [[MSAuthTokenInfo alloc] initWithAuthToken:authTokenInfo.authToken
+                                         andStartTime:authTokenInfo.startTime
+                                           andEndTime:nextChangeTime];
+  }
+  return authTokenInfo;
+   */
+}
+
+- (void)saveAuthToken:(nullable NSString *)authToken withAccountId:(nullable NSString *)accountId expiresOn:(nullable NSDate *)expiresOn {
+  @synchronized(self) {
+
+    // Read token array from storage.
+    NSMutableArray<MSAuthTokenInfo *> *authTokensHistory = [self authTokensHistoryState];
+    NSString *latestAuthToken = [authTokensHistory lastObject].authToken;
+    if ((latestAuthToken != authToken) &&
+        !(latestAuthToken && authToken && [latestAuthToken isEqualToString:(NSString * _Nonnull) authToken])) {
```

Comment:
```suggestion
        ![latestAuthToken isEqualToString:authToken]) {
```

Reply:
Unfortunately, your suggestion doesn't work. There are 2 issues since both `latestAuthToken` and `authToken` are nullables:
1. We also need to check that `latestAuthToken != nil` to avoid situations like this: `[nil isEqualToString:nil] == NO`.
2. Compilation fails if we're passing nullable argument to `isEqualToString`: `Error:(52, 61) implicit conversion from nullable pointer 'NSString * _Nullable' to non-nullable pointer type 'NSString * _Nonnull'`.

But I agree that this check isn't really clear so I did some refactoring and simplified it using a ternary operator. I hope it looks better now. Thanks.